src/Npgsql/NpgsqlDataSource.cs

```diff
@@ -171,6 +195,42 @@ public static NpgsqlDataSource Create(string connectionString)
     public static NpgsqlDataSource Create(NpgsqlConnectionStringBuilder connectionStringBuilder)
         => Create(connectionStringBuilder.ToString());
 
+    internal async Task Bootstrap(NpgsqlConnector connector, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken)
+    {
+        var hasSemaphore = async
```

Comment:
This seems weird. Since we can call it while opening the connector, if there are multiple connectors at the same time, every single one of them is going to load the database info (even if sequentially). We probably should add a flag `force`, and only go forward after we've got the semaphore if it's not bootstrapped and the flag is set.

Reply:
Maybe even make the flag `IsBootstrapped` private and check it only here...

Reply:
Good catch.